vim-patch:8.1.{1769, 1772, 1791},8.2.{1747}

skippi opened the pull request:
Problem:    'shellslash' is also used for completion.
Solution:   Add the 'completeslash' option. (Yasuhiro Matsumoto, closes vim/vim#3612)
https://github.com/vim/vim/commit/ac3150d385e6e3f3fe76642aac3cda954d30583f

skippi committed:
vim-patch:8.1.1769: 'shellslash' is also used for completion

Problem:    'shellslash' is also used for completion.
Solution:   Add the 'completeslash' option. (Yasuhiro Matsumoto, closes vim/vim#3612)
https://github.com/vim/vim/commit/ac3150d385e6e3f3fe76642aac3cda954d30583f

skippi commented on src/nvim/option_defs.h:
```diff
@@ -743,6 +746,7 @@ enum {
   , BV_CPT
   , BV_DICT
   , BV_TSR
+  , BV_CSL
```
This change isn't wrapped in an `#ifdef` because it prevents the generated option header from finding the `BV_CSL` symbol. It's a bit of a hack, so let me know if there's a more ideal resolution.

janlazo replied:
I don't see any other symbol with an `#ifdef` so this is ok.

skippi commented on runtime/doc/options.txt:
```diff
@@ -1386,6 +1386,21 @@ A jump table for the options with a short description can be found at |Q_op|.
+						*'completeslash'* *'csl'*
+'completeslash' 'csl'	string	(default: "")
+			local to buffer
+			{not in Vi}  {only for MS-Windows}
```
For what it's worth, the rest of the options list "Windows" instead of "MS-Windows". Is there a precedent for fixing these sorts of things when porting patches?

skippi replied:
Looks like this is already handled in a different patch, see [here](https://github.com/neovim/neovim/pull/13116#discussion_r508112107).

janlazo commented on runtime/doc/options.txt:
```diff
@@ -1386,6 +1386,21 @@ A jump table for the options with a short description can be found at |Q_op|.
+	  completion in insert mode. This is useful when editing HTML tag, or
+	  Makefile with 'noshellslash' on Windows.
+	- When this option is set to "backslash", backslash is used. This is
+	  useful when editing a batch file with 'shellslash' set on Windows.
```
Patch 8.2.0017, https://github.com/vim/vim/commit/6f345a14, replaces `Windows` with `MS-Windows`.

skippi replied:
In retrospect, I think I'll keep this out of the PR, since this looks like a good intro patch for other contributors. Also would increase the scope of the PR beyond `completeslash`.

janlazo commented:
```sh
> ./scripts/vim-patch.sh -L --grep completeslash
v8.1.1769
v8.1.1772
v8.1.1791
v8.2.1747
```

`completeslash` patches go together. If a patch fails, look for a newer patch that fixes it.

janlazo commented on src/nvim/generators/gen_options.lua:
```diff
@@ -144,6 +144,9 @@ local dump_option = function(i, o)
   if o.enable_if then
     w('#endif')
   end
+  if o.enable_if then
```
Does ASAN error after patch 8.1.1772 without the changes here?

skippi replied:
It looks like it still does. Related snippet from running ASAN locally:
```bash
> CC=clang make CMAKE_FLAGS="-DCLANG_ASAN_UBSAN=ON"
...
FAILED: runtime/doc/tags
cd /mnt/c/src/neovim/build/runtime && /usr/bin/cmake -E remove doc/* && /usr/bin/cmake -E copy_directory
/mnt/c/src/neovim/runtime/doc doc && /mnt/c/src/neovim/build/bin/nvim -u NONE -i NONE -e --headless -c he
lptags\ ++t\ doc -c quit
../src/nvim/option.c:1025:26: runtime error: load of null pointer of type 'char_u *' (aka 'unsigned char
*')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/nvim/option.c:1025:26 in
```

janlazo replied:
Your fix is correct. `gen_options.lua` is not doing enough to handle the cases when `enable_if` is defined. I see what I can do to reduce `enable_if` and add default values for for struct members in `#else`.

Ideal case is that adding the condition from  `enable_if` is done once.

skippi replied:
I'll consider my end of the PR complete then. If there's anything else left for me to do, send me a holler.

janlazo replied:
Rebase on master because of my changes and then I'll review.

skippi commented:
I'll work on incorporating these patches into the PR. I'll give an update on ASAN afterwards.

skippi committed:
vim-patch:8.2.1747: result of expand() unexpectedly depends on 'completeslash'

Problem:    Result of expand() unexpectedly depends on 'completeslash'.
Solution:   Temporarily reset 'completeslash'. (Yasuhiro Matsumoto,
            closes vim/vim#7021)
https://github.com/vim/vim/commit/8f187fc6304222956f94a700758a490cc8c0af99

janlazo commented on src/nvim/edit.c:
```diff
@@ -4171,6 +4171,22 @@ static int ins_compl_get_exp(pos_T *ini)
+              } else if (curbuf->b_p_csl[0] == 'b' && *ptr == '/') {
+                *ptr = '\\';
+              }
+              ptr += (*mb_ptr2len)(ptr);
```
New code should use `utfc_ptr2len` instead of the macro.

janlazo commented on src/nvim/ex_getln.c:
```diff
@@ -5052,6 +5052,22 @@ ExpandFromContext (
+          } else if (p_csl[0] == 'b' && *ptr == '/') {
+            *ptr = '\\';
+          }
+          ptr += (*mb_ptr2len)(ptr);
```
https://github.com/neovim/neovim/pull/13116/files?file-filters%5B%5D=.c&file-filters%5B%5D=.h#r511653780

janlazo commented on src/nvim/ex_getln.c:
```diff
@@ -5052,6 +5052,22 @@ ExpandFromContext (
+#ifdef BACKSLASH_IN_FILENAME
+    if (p_csl[0] != NUL && (options & WILD_IGNORE_COMPLETESLASH) == 0) {
+      int i;
+      for (i = 0; i < *num_file; i++) {
```
https://neovim.io/develop/style-guide.xml?showone=Local_Variables#Local_Variables

janlazo reviewed (changes requested):
LGTM aside from style issues.

janlazo commented:
Patch 8.1.1791 disappeared :confused:

skippi committed:
vim-patch:8.1.1772: options test still fails on MS-Windows

Problem:    Options test still fails on MS-Windows.
Solution:   Check buffer-local value of 'completeslash'.
https://github.com/vim/vim/commit/b78564d0221089e6dfc9c9d58239c18b991ca9fe

skippi committed:
vim-patch:8.1.1791: 'completeslash' also applies to globpath()

Problem:    'completeslash' also applies to globpath().
Solution:   Add the WILD_IGNORE_COMPLETESLASH flag. (test by Yasuhiro
            Matsumoto, closes vim/vim#4760)

janlazo commented on src/nvim/edit.c:
```diff
@@ -4171,6 +4171,22 @@ static int ins_compl_get_exp(pos_T *ini)
+              } else if (curbuf->b_p_csl[0] == 'b' && *ptr == '/') {
+                *ptr = '\\';
+              }
+              ptr += (*utfc_ptr2len)(ptr);
```
I don't understand. This is an actual function, not a macro. Why wrap in `(* )`?

skippi replied:
Bad on my part. That was a blatant copy paste without thinking about the surrounding syntax.

janlazo commented on src/nvim/edit.c:
```diff
@@ -4171,6 +4171,22 @@ static int ins_compl_get_exp(pos_T *ini)
+#ifdef BACKSLASH_IN_FILENAME
+        if (curbuf->b_p_csl[0] != NUL) {
+          int i;
+          for (i = 0; i < num_matches; i++) {
```
`int i = 0;`. I don't see `i` used outside of the for loop.

janlazo reviewed (approved):
LGTM
